url0d38a10ee7_diegopliebana_GEBT_tgz-pJ8-ch_idsia_scenarios_test_EvolveWithChangingSeedsJ8/src/grammaticalbehaviors/GEBT_Mario/Graph.java

```diff
@@ -0,0 +1,1289 @@
+        long id = (100000000L * a_aID) + a_bID;
+        Long idKey = new Long(id);
+
+        if(!m_edges.containsKey(idKey))
+        {
+            //No edge
+            return null;
+        }
+
+        return m_edges.get(idKey);
+    }
+
+    //Gets and edge betwwen two nodes.
+    public Edge getEdge(long a_ID)
+    {
+        if(!m_edges.containsKey(a_ID))
+        {
+            //No edge
+            return null;
+        }
+
+        return m_edges.get(a_ID);
+    }
+
+
+    //Removes node and edges to and from it.
+    public void removeNode(int a_x, int a_y)
+    {
+        //TO DO.
+    }
+
+
+    private int nodeGetX(int a_id)
+    {
+        return a_id % 10000;
+    }
+
+
+    private int nodeGetY(int a_id)
+    {
+        int chop = (int) (a_id * 0.00001);
+        return chop % 100;
+    }
+
+    public Object[] getNodes()
+    {
+        return m_nodes.values().toArray();
+    }
+
+    public Object[] getEdges()
+    {
+        return m_edges.values().toArray();
+    }
+
+
+    public Node getNode(int a_id)
+    {
+        return m_nodes.get(a_id);
+    }
+
+    public int getNumNodes() {return m_nodes.size();}
+
+
+    public int getRightMostNode(Vector<Integer> a_forbidden)
+    {
+        //nodes
+        Object[] coll = m_nodes.values().toArray();
+        int maxX = 0;
+        int maxXNode = -1;
+        for(int i = 0; i < m_nodeCounter; ++i)
+        {
+            Node n = (Node)coll[i];
+
+            //1st, it is not forbidden.
+            if(!a_forbidden.contains(n.getID()))
+            {
+                //2nd, check distance.
+                if(n.getX() > maxX)
+                {
+                    maxX = n.getX();
+                    maxXNode = n.getID();
+                }
+            }
+        }
+
+        return maxXNode;
+    }
+
+    public int getLeftMostNodeClose(int a_originNode, Vector<Integer> a_forbidden)
+    {
+        //nodes
+        Object[] coll = m_nodes.values().toArray();
+        int minX = Integer.MAX_VALUE;
+        int minXNode = -1;
+        int minDistance = Integer.MAX_VALUE;
+        Node orgNode = m_nodes.get(a_originNode);
+
+        for(int i = 0; i < m_nodeCounter; ++i)
+        {
+            Node n = (Node)coll[i];
+            //1st, it is not forbidden.
+            if(!a_forbidden.contains(n.getID()))
+            {
+                int distance = n.manhattanDistanceTo(orgNode);
+
+                //2nd, check distance.
+                if(n.getX() < minX)
+                {
+                    minX = n.getX();
+                    minXNode = n.getID();
+                    minDistance = distance;
+                }
+                else if(n.getX() == minX)
+                {
+                    if(distance < minDistance)
+                    {
+                        minX = n.getX();
+                        minXNode = n.getID();
+                        minDistance = distance;
+                    }
+                }
+            }
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 500:525, 539:566, 580:607, 622:649.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.